signimage/signimage.go

```diff
@@ -0,0 +1,352 @@
+	flag.StringVar(&pullsecret, "pushsecret", "", "path to file containing credentials for pushing images")
+	flag.BoolVar(&nopush, "no-push", false, "do not push the resulting image")
+
+	flag.Parse()
+
+	checkarg(&unsignedimagename, "UNSIGNEDIMAGE", "")
+	checkarg(&signedimagename, "SIGNEDIMAGE", unsignedimagename+"signed")
+	checkarg(&fileslist, "FILESTOSIGN", "")
+	checkarg(&privkeyfile, "KEYSECRET", "")
+	checkarg(&pubkeyfile, "CERTSECRET", "")
+	checkarg(&pullsecret, "PULLSECRET", "")
+	checkarg(&pushsecret, "PUSHSECRET", pullsecret)
+	// if we've made it this far the arguements are sane
+
+	// get a temp dir to copy kmods into for signing
+	extractiondir, err := os.MkdirTemp("/tmp/", "kmod_signer")
+	if err != nil {
+		die(1, "could not create temp dir: %w", err)
+	}
+	defer os.Remove(extractiondir)
+
+	// sets up a tar archive we will use for a new layer
+	var b bytes.Buffer
+	tarwriter := tar.NewWriter(&b)
+
+	//make a map of the files to sign so we can track what we want to sign
+	kmodstosign := make(map[string]string)
+	for _, x := range strings.Split(fileslist, ":") {
+		kmodstosign[x] = "not found"
+	}
+
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(unsignedimagename, opts...)
+	if err != nil {
+		die(1, "could not parse the container image name", err)
+	}
+
+	a, err := getAuthFromFile(pullsecret, strings.Split(unsignedimagename, "/")[0])
+	if err != nil {
+		die(1, "failed to get auth", err)
+	}
+
+	descriptor, err := remote.Get(ref, remote.WithAuth(a))
+	if err != nil {
+		te := &transport.Error{}
+
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			die(1, "could not get image http.StatusNotFound", err)
+		}
+
+		die(1, "could not get image", err)
+	}
+
+	img, err := descriptor.Image()
+	if err != nil {
+		die(1, "could not Image()", err)
+	}
+	fmt.Printf("\n== Successfully pulled image: %s\n", unsignedimagename)
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
+
+	/*
+	** loop through all the layers in the image from the top down
+	 */
+	layers, err := img.Layers()
+	if err != nil {
+		die(127, "could not get the layers from the fetched image", err)
+	}
+	for i := len(layers) - 1; i >= 0; i-- {
+		fmt.Printf("== Searching layer %d\n", i)
+		currentlayer := layers[i]
+		layerreader, err := currentlayer.Uncompressed()
+		if err != nil {
+			die(1, "could not get layer", err)
+		}
+
+		/*
+		** loop through all the files in the layer
+		 */
+		tarreader := tar.NewReader(layerreader)
+		for {
+			header, err := tarreader.Next()
+			if err == io.EOF || header == nil {
+				break // End of archive
+			}
+
+			// paths in a layer are relative, and supplied paths are absolute so canonicalise
+			canonfilename := canonicalisePath(header.Name)
+			//either the kmod has not yet been found, or we didn't define a list to search for
+			if kmodstosign[canonfilename] == "not found" || (fileslist == "" && kmodstosign[canonfilename] == "") {
```

Comment:
i might be missing something, but it seems that in case of an empty filelist, this code will be signing every file/directory in the layer,no?

Reply:
yep, that was a bug that crept in while I was debugging something else :)  should be fixed now